File: project.py
from pyspark import SparkContext
from pyspark.sql import SQLContext
from pyspark.sql.functions import col
import matplotlib.pyplot as plt
from itertools import combinations
import itertools
import numpy as np
from pprint import pprint
from itertools import chain
from operator import add
from math import sqrt
from pyspark.mllib.clustering import KMeans, KMeansModel
from pyspark.mllib.clustering import GaussianMixture, GaussianMixtureModel

import matplotlib.pyplot as plt

sc = SparkContext()
ratingsFile = sc.textFile('ratings_Video_Games.10k.csv')

# returns (user, item, rating, time)
ratings = ratingsFile.map(lambda line: (line.split(','))).cache()

#check ratings distribution
# ratings_dist = ratings.map(lambda x: (x[2], x[1])).map(lambda x: (x[0], [x[1]])).reduceByKey(lambda x, y: x + y)\
#     .map(lambda x: (x[0], len(x[1]))).sortBy(lambda x: x[0]).collect()
# pprint(ratings_dist)
# item = zip(*ratings_dist)[0]
# count = zip(*ratings_dist)[1]
# x_pos = np.arange(len(item))
# slope, intercept = np.polyfit(x_pos, count, 1)
# plt.bar(x_pos, count,align='center')
# plt.xticks(x_pos, item)
# plt.ylabel('Number of reviews')
# plt.xlabel('Rating')
# plt.show()

item_user = ratings.map(lambda x: (x[1], x[0])).map(lambda x: (x[0], [x[1]])).reduceByKey(lambda x, y: x + y)
item_popularity = item_user.map(lambda x: (x[0], len(x[1]))).sortBy(lambda x: -x[1]).collect()
max_ratings = item_popularity[0][1]
print max_ratings

def findItemFeatures(x, longtail_thrs, max_ratings):
    rating_sum = 0.0
    for rating in x[1]:
        rating_sum += float(rating)
    rating_mean = float(rating_sum) / len(x[1])
    if(len(x[1]) > longtail_thrs*max_ratings):
        return (x[0], (rating_mean, float(len(x[1])), 'HEAD'))
    else:
        return (x[0], (rating_mean, float(len(x[1])), 'TAIL'))

longtail_thrs = 0.05
item_ftrs_map = ratings.map(lambda x:(x[1], x[2])).map(lambda x: (x[0], [x[1]])).reduceByKey(lambda x, y: x + y)\
    .map(lambda x: findItemFeatures(x, longtail_thrs, max_ratings)).sortBy(lambda x: -x[1][1])

pprint(item_ftrs_map.mapValues(list).collect())
head_ratings_dist = item_ftrs_map.filter(lambda x:x[1][2] == 'TAIL').map(lambda x: (x[1][0], x[1][1]))
pprint(head_ratings_dist.collect())


pprint(item_ftrs_map.collect())

# Clustering
# item_ftrs = item_ftrs_map.map(lambda x: np.array([x[1][0], x[1][1]]))
# # item_ftrs = item_ftrs_map.map(lambda x: np.array(x[1]))
# model = KMeans.train(item_ftrs, 10)
# print("Final centers: " + str(model.clusterCenters))
# print("Total Cost: " + str(model.computeCost(item_ftrs)))

# pprint(item_mean_rating.collect())
#
#
# # # # plot longtail with matplotlib after bringing RDD to local
# # # item = zip(*item_popularity)[0]
# # # count = zip(*item_popularity)[1]
# # # x_pos = np.arange(len(item))
# # # slope, intercept = np.polyfit(x_pos, count, 1)
# # # plt.bar(x_pos, count,align='center')
# # # plt.xticks(x_pos, item)
# # # plt.ylabel('Popularity')
# # # plt.xlabel('Items')
# # # plt.show()

File: project.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.Row
import org.apache.spark.mllib.feature.{HashingTF, IDF}
import org.apache.spark.mllib.linalg.Vector
import org.apache.spark.rdd.RDD


/**
  * Created by dexter on 12/16/2016.
  */
object project {

  def getHeadTail( a:Int, longtail_thrs:Double, max_ratings:Int ): String = {
    if(a > longtail_thrs*max_ratings) {
      "HEAD"
    }else{
      "TAIL"
    }
  }

  def main(args: Array[String]) {
    val ratingsFile = "ratings_Video_Games.10k.csv"
    val metadataFile = "meta_Video_Games.json"
    val conf = new SparkConf().setMaster("local[2]").setAppName("Hello Spark")
    val sc = new SparkContext(conf)
    val ss = SparkSession.builder.getOrCreate()

    val ratings = sc.textFile(ratingsFile).map(line => (line.split(','))).cache()
    val item_user = ratings.map(x => (x(1), x(0))).groupByKey().mapValues(_.toList)
    val item_popularity = item_user.map(x => (x._1, x._2.size)).sortBy(x => -x._2).collect()

    val max_ratings = item_popularity(0)._2
    val longtail_thrs = 0.05
    val ratingsRdd = ratings.map(x => (x(1), x(2))).groupByKey().map(x => (x._1, List(getHeadTail(x._2.size, longtail_thrs, max_ratings), x._2.size)))

    val metadataDF = ss.read.json(metadataFile)

    metadataDF.createOrReplaceTempView("metadata_table")
    val items = ss.sql("Select asin, categories from metadata_table ").na.drop()
    val itemsRdd = items.rdd.map(x => (x(0).toString, List(x(1))))
    val itemNewRDD = (ratingsRdd union itemsRdd).reduceByKey(_ ++ _)
    val categories = itemNewRDD.map(x => x._2)

    val hashingTF = new HashingTF()
    val tf: RDD[Vector] = hashingTF.transform(categories)
    tf.cache()
    val idf = new IDF().fit(tf)
    val tfidf: RDD[Vector] = idf.transform(tf)
    val idfIgnore = new IDF(minDocFreq = 2).fit(tf)
    val tfidfIgnore: RDD[Vector] = idfIgnore.transform(tf)

    itemsRdd.collect().foreach(println)
    ratingsRdd.collect().foreach(println)
    itemNewRDD.collect().foreach(println)
    categories.collect().foreach(println)


    println("END")
  }

}
